### PR TITLE
Determine hardware capabilities, phase 1

### PR DESCRIPTION
**Pull request by DethRaid**
Adds in basic hardware capability querying

The plan is to add in more things are we need them. This is mostly getting the device vendor and some info about which Turing features are available

### CODE REVIEW

File: src/render_engine/gl3/gl3_render_engine.cpp [hunk 1]
```diff
@@ -11,18 +12,52 @@
...
         swapchain = new Gl3Swapchain(settings.settings.max_in_flight_frames, window->get_window_size());
 
         set_initial_state();
     }
 
     Gl4NvRenderEngine::~Gl4NvRenderEngine() { delete swapchain; }
 
+    void Gl4NvRenderEngine::save_device_info() {
+        const GLubyte* raw_vendor_string = glGetString(GL_VENDOR);
+        std::string vendor_string(reinterpret_cast<const char*>(raw_vendor_string));
```
Comment (cwfitzgerald) [inline src/render_engine/gl3/gl3_render_engine.cpp:30]: iirc static cast should work for this

File: src/render_engine/gl3/gl3_render_engine.cpp [hunk 2]
```diff
@@ -11,18 +12,52 @@
...
+        std::transform(vendor_string.begin(), vendor_string.end(), vendor_string.begin(), [](const char c) { return std::tolower(c); });
```
Comment (cwfitzgerald) [inline src/render_engine/gl3/gl3_render_engine.cpp:31]: ugh this is why c++ is dumb

File: src/render_engine/gl3/gl3_render_engine.cpp [hunk 3]
```diff
@@ -38,7 +73,7 @@ namespace nova::renderer::rhi {
         glEnableVertexAttribArray(6);
     }
 
-    void Gl4NvRenderEngine::set_num_renderpasses([[maybe_unused]] uint32_t num_renderpasses) {
+    void Gl4NvRenderEngine::set_num_renderpasses(uint32_t /* num_renderpasses */) {
```
Comment (cwfitzgerald) [inline src/render_engine/gl3/gl3_render_engine.cpp:76]: why the change?

File: src/render_engine/vulkan/vulkan_render_engine.cpp [hunk 4]
```diff
@@ -1208,6 +1211,56 @@ namespace nova::renderer::rhi {
...
+            return std::strcmp(ext_name, ext_props.extensionName) == 0;
+        };
+
+        // TODO: use std::bind_front instead of std::bind when C++20 drops
+        using namespace std::placeholders;
+
+        // TODO: Update as more GPUs support hardware raytracing
+        const auto rt_ext_itr = std::find_if(available_extensions.begin(),
+                                             available_extensions.end(),
+                                             std::bind(extension_name_matcher, VK_NV_RAY_TRACING_EXTENSION_NAME, _1));
```
Comment (cwfitzgerald) [inline src/render_engine/vulkan/vulkan_render_engine.cpp]: bind considered harmful

File: src/nova_renderer.cpp [hunk 5]
```diff
@@ -497,7 +497,7 @@ namespace nova::renderer {
 
         for(const auto& [descriptor_name, resource_name] : bindings) {
             const rhi::ResourceBindingDescription& binding_desc = descriptor_descriptions.at(descriptor_name);
```
Comment (cwfitzgerald) [inline src/nova_renderer.cpp:499]: A+ for using `at`

File: src/render_engine/gl3/glad.c [hunk 6]
```diff
@@ -1,32 +1,35 @@
 /*
```
Comment (cwfitzgerald) [inline src/render_engine/gl3/glad.c:1]: Why do we have glad in tree at all/why isn't this in external?

**Review by cwfitzgerald — CHANGES_REQUESTED** [review]
Some minor nits.

File: src/render_engine/gl3/gl3_render_engine.cpp (hunk 1 above)
Comment (DethRaid) [reply src/render_engine/gl3/gl3_render_engine.cpp:30]: MSVC disagrees :(

File: src/render_engine/gl3/gl3_render_engine.cpp (hunk 2 above)
Comment (DethRaid) [reply src/render_engine/gl3/gl3_render_engine.cpp:31]: Yeah I had a whole rant about this at one point

File: src/render_engine/gl3/gl3_render_engine.cpp (hunk 3 above)
Comment (DethRaid) [reply src/render_engine/gl3/gl3_render_engine.cpp:76]: _someone_ told me to not use `[[maybe_unused]]` for unused method parameters and instead to comment out their names

File: src/nova_renderer.cpp (hunk 5 above)
Comment (DethRaid) [reply src/nova_renderer.cpp:499]: I always use at

File: src/render_engine/gl3/gl3_render_engine.cpp (hunk 2 above)
Comment (cwfitzgerald) [reply src/render_engine/gl3/gl3_render_engine.cpp:31]: I remember seeing it lmao, I wish we could backport some of the rust iterator interface to c++.

File: src/render_engine/gl3/glad.c (hunk 6 above)
Comment (DethRaid) [reply src/render_engine/gl3/glad.c:1]: Because GLAD

GLAD is meant for you to include it in the tree

If you want to wrap it up in a CMake and PR it in that's fine but I have no desire at all to do any of that

File: src/render_engine/gl3/gl3_render_engine.cpp (hunk 3 above)
Comment (cwfitzgerald) [reply src/render_engine/gl3/gl3_render_engine.cpp:76]: Will this never use the parameter?

File: src/render_engine/gl3/gl3_render_engine.cpp (hunk 2 above)
Comment (DethRaid) [reply src/render_engine/gl3/gl3_render_engine.cpp:31]: Whenever I get around to making a Nova Template Library it'll have something like Java Streams or C# LINQ or Rust iter()

File: src/render_engine/gl3/glad.c (hunk 6 above)
Comment (cwfitzgerald) [reply src/render_engine/gl3/glad.c:1]: lmao, ugh, if I care I'll fix it tmmr.

File: src/render_engine/gl3/gl3_render_engine.cpp (hunk 3 above)
Comment (DethRaid) [reply src/render_engine/gl3/gl3_render_engine.cpp:76]: The OpenGL backend doesn't. This method lets the Vulkan and D3D12 backends allocate things for their renderpasses. OpenGL has no notion of renderpasses and thus doesn't need to allocate anything for them

File: src/render_engine/vulkan/vulkan_render_engine.cpp (hunk 4 above)
Comment (DethRaid) [reply src/render_engine/vulkan/vulkan_render_engine.cpp]: C++20 hasn't given us `std::bind_front` yet so idk. What should I use instead?

File: src/render_engine/gl3/gl3_render_engine.cpp (hunk 3 above)
Comment (cwfitzgerald) [reply src/render_engine/gl3/gl3_render_engine.cpp:76]: :+1:

File: src/render_engine/vulkan/vulkan_render_engine.cpp (hunk 4 above)
Comment (cwfitzgerald) [reply src/render_engine/vulkan/vulkan_render_engine.cpp]: A lambda? 

```cpp
[](auto x) { return extension_name_matcher(VK_NV_RAY_TRACING_EXTENSION_NAME, x); }
```
or something similar

Comment (DethRaid) [reply src/render_engine/vulkan/vulkan_render_engine.cpp]: I hate it but I hate everything so it's fine

**Review by cwfitzgerald — APPROVED** [review]
LGTM
